### PR TITLE
Improved zsh completion time from ~0.5s to near instant

### PR DESCRIPTION
**Pull request by racko**
The "catkin locate -s" takes around 0.5s. It gets called whenever the cache is checked.

### CODE REVIEW

File: completion/_catkin [hunk 1]
```diff
@@ -102,7 +102,8 @@ _catkin_packages_caching_policy() {
   if [[ -z "$_workspace_source_space" ]]; then
     _debug_log "Searching for source space..."
 
-    _workspace_source_space="$(catkin locate -s --quiet)"
+    _catkin_get_enclosing_workspace $_workspace_root_hint
+    _workspace_source_space="$_workspace_root/src"
```
Comment (wjwwood) [inline completion/_catkin]: This is a flawed assumption, the source directory is not always `<workspace dir>/src`. Since it can be any folder, that's why `catkin locate` is required to look up the setting.

**Comment by racko** [comment]
**tl;dr:** No idea how to solve this "properly". What about adding a special case for "$_workspace_root/src" since it's what works for 99% of all users? It is the recommended layout after all ...

---

Well ... 
```
import catkin_tools.context
ctx = catkin_tools.context.Context.load('/home/racko/catkin_ws', None, type('Options', (object,), dict(space='src'))(), load_env=False)
print ctx.source_space_abs
```
takes around 250ms.

```
from catkin_tools.metadata import get_metadata, get_active_profile
print get_metadata('/home/racko/catkin_ws', get_active_profile('/home/racko/catkin_ws'), 'config')['source_space']
```
takes around 175ms. Calling `get_metadata` or `get_active_profile` alone also takes around 175ms, so among other things, we can conclude that calling `get_active_profile` and then reading the config.yaml with another tool will not be good enough. But still it is interesting to measure.

```
import yaml
print yaml.load(open('/home/racko/catkin_ws/.catkin_tools/profiles/default/config.yaml'))['source_space']
```
takes around 30ms.

```
yaml2json - < /home/racko/catkin_ws/.catkin_tools/profiles/default/config.yaml | jq '.source_space'
```
takes around 20ms.

Also caching the source space path is not an option: To check if the cache is invalid, we need to check the date of the config.yaml, and getting the path takes around 175ms again ...

Reimplementing the get_active_profile logic in zsh (or some compiled language) would be trivial and fast enough, but duplicating the logic is of course not an option. And I'm certain that implementing the logic in C++ and using it in python (similar to tf) was decided against long before I came along, besides it being to big of a change for this pull request :)

**Comment by wjwwood** [comment]
If the issue is only due to asking for the version number, perhaps we could delay acquiring that information until it is asked for.

However, I think in general the entry points performance is a bigger issue, see: https://github.com/catkin/catkin_tools/issues/297

**Comment by racko** [comment]
I close this pull request because the proposed solution isn't actually working:


> What about adding a special case for "$_workspace_root/src" since it's what works for 99% of all users? It is the recommended layout after all ...
> 
> Not even this special case works: Finding "$_workspace_root/src" does not mean that it is the source space :(

Also my suggestion that replacing `pkg_resources.require("catkin_tools")[0].version` would fix the problem seems to be false: While it certainly speeds up migrate_metadata, I couldn't observe a speed-up for `catkin locate -s` or the zsh auto-complete. I guess that `pkg_resources.require("catkin_tools")` or something similar gets called anyhow.
